### ISSUE TITLE
return-undefined false positive

### ISSUE THREAD

**Issue by cartant** [palantir/tslint#2812]
### Bug Report

- __TSLint version__: 5.3.0
- __TypeScript version__: 2.3.3
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
export function t(p: (a: any) => any): void {}
t((a) => { return undefined; });
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "return-undefined": { "severity": "error" }
  }
}

```

#### Actual behavior

```
ERROR: ...: `void` function should use `return;`, not `return undefined;`.
```

#### Expected behavior

No error, as the function parameter is typed as returning `any` - not `void`.

Prior to updating to the latest version of TSLint, the configuration did not see an error effected. No error is effected if 5.2.0 is used.

**Comment by adidahiya** [comment]
looks like a direct result of #2731. notice the test baseline that got changed: https://github.com/palantir/tslint/pull/2731/files#diff-bd90f0718fe7a8725e9db572ef333339R36

can you provide a bit more context around the code your'e trying to lint here? should it really be typed `any`?

... either way, this behavior is kind of surprising, so we should probably revert that part of the rule change @andy-hanson

**Comment by cartant** [comment]
@adidahiya Sure. The typings aren't mine and `any` is reasonably appropriate for the usage.

It's a [`transactionUpdate` parameter](https://github.com/firebase/firebase-js-sdk/blob/v4.0.0/typings/firebase.d.ts#L291) in the Firebase JavaScript client:

```ts
transaction (
  transactionUpdate : (a : any ) => any ,
  onComplete ? : (a : Error | null , b : boolean ,
  c : firebase.database.DataSnapshot | null
) => any , applyLocally ? : boolean ) : firebase.Promise < any > ;
```

I need to implement that interface in a mock.

**Comment by cartant** [comment]
@adidahiya And the code the fails the linting is [the test for the mock](https://github.com/cartant/firebase-nightlight/blob/v2.0.2/source/mock-ref-spec.ts#L1673-L1676):

```ts
return mockRef
    .transaction((value) => {

        return undefined;
    })
    .then(({ committed, snapshot }) => {
```

Firebase transactions return `undefined` if nothing is updated.

**Comment by cartant** [comment]
Seeing as though this is a breaking change, is it going to be considered a bug? Or should I implement the trivial workaround that'll be required if this is indeed how you want the rule to work?

**Comment by adidahiya** [comment]
I consider it a bug. Will label it as such.
